File: surgical_replace.js

const fs = require('fs');
const path = require('path');

function surgicalReplace(filePath, startLine, endLine, newContentFile) {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');

    const newContent = fs.readFileSync(newContentFile, 'utf-8');

    // startLine and endLine are 1-indexed
    const before = lines.slice(0, startLine - 1);
    const after = lines.slice(endLine);

    const result = before.join('\n') + '\n' + newContent + '\n' + after.join('\n');

    fs.writeFileSync(filePath, result, 'utf-8');
    console.log('Replacement complete.');
}

const args = process.argv.slice(2);
if (args.length !== 4) {
    console.log('Usage: node surgical_replace.js <file_path> <start_line> <end_line> <new_content_file>');
    process.exit(1);
}

surgicalReplace(args[0], parseInt(args[1]), parseInt(args[2]), args[3]);

File: surgical_replace.py

import sys

def surgical_replace(file_path, start_line, end_line, new_content_file):
    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    with open(new_content_file, 'r', encoding='utf-8') as f:
        new_content = f.read()
    
    # Lines are 1-indexed, so start_line-1 is the first line to replace
    # end_line is inclusive
    before = lines[:start_line-1]
    after = lines[end_line:]
    
    with open(file_path, 'w', encoding='utf-8') as f:
        f.writelines(before)
        f.write(new_content)
        f.writelines(after)

if __name__ == "__main__":
    if len(sys.argv) != 5:
        print("Usage: python surgical_replace.py <file_path> <start_line> <end_line> <new_content_file>")
        sys.exit(1)
    
    surgical_replace(sys.argv[1], int(sys.argv[2]), int(sys.argv[3]), sys.argv[4])
    print("Replacement complete.")
